c2corg_api/views/document.py

```diff
@@ -493,6 +506,53 @@ def _check_versions(self, document, document_in):
                 raise HTTPConflict('version of geometry has changed')
 
 
+def validate_geometry_type(document):
+    """ Checks that geometry field contains valid geometry for each doc. type:
+        articles, books - no geometry
+        images, xreports, user_profiles, waypoints - should be only Point
+        areas, maps - should be only Polygon
+        outings, routes - should be only LineString or MultiLineString
+        If there is no geometry, nothing has to be validated.
+    :param document:
+    :return:
+    """
+    doc_type = document.type
+
+    if doc_type in [OUTING_TYPE, ROUTE_TYPE, AREA_TYPE, MAP_TYPE]:
+        g_field = str(document.geometry.geom_detail)
+    elif doc_type in [IMAGE_TYPE, USERPROFILE_TYPE,
+                      XREPORT_TYPE, WAYPOINT_TYPE]:
+        g_field = str(document.geometry.geom)
+
+    if g_field == 'None':
+        return True
+
+    try:
+        s = wkb.loads(g_field, hex=True)
+        geom_type = s.type
+    except Exception as e:
+        logging.exception(e)
+        try:
+            q = "SELECT GeometryType(ST_GeomFromEWKT('" + g_field + "'));"
```

Comment:
Instead of making a request to the database, couldn't you parse the WKB to GeoJSON and then check the `type` property? See https://github.com/c2corg/v6_api/blob/master/c2corg_api/ext/colander_ext.py#L69

And also: http://norvig.com/21-days.html#answers